extensions/GitExtension/FileExplorerGitIntegration/Models/StatusCache.cs

```diff
@@ -0,0 +1,329 @@
+        }
+
+        // Diff old and new status to obtain a list of files to refresh to the Shell.
+        if (oldStatus == null)
+        {
+            return;
+        }
+
+        HashSet<string> changed = [];
+        foreach (var newEntry in newStatus.Entries)
+        {
+            GitStatusEntry? oldValue;
+            if (oldStatus.Entries.TryGetValue(newEntry.Key, out oldValue))
+            {
+                if (newEntry.Value.Status != oldValue.Status)
+                {
+                    changed.Add(newEntry.Key);
+                }
+
+                oldStatus.Entries.Remove(newEntry.Key);
+            }
+            else
+            {
+                changed.Add(newEntry.Key);
+            }
+        }
+
+        foreach (var oldEntry in oldStatus.Entries)
+        {
+            changed.Add(oldEntry.Key);
+        }
+
+        foreach (var entry in changed)
+        {
+            var fixedPath = Path.Combine(_workingDirectory, entry).Replace(Path.AltDirectorySeparatorChar, Path.DirectorySeparatorChar);
+            unsafe
+            {
+                IntPtr strPtr = Marshal.StringToCoTaskMemUni(fixedPath);
+                PInvoke.SHChangeNotify(Windows.Win32.UI.Shell.SHCNE_ID.SHCNE_UPDATEITEM, Windows.Win32.UI.Shell.SHCNF_FLAGS.SHCNF_PATH, (void*)strPtr, null);
+                Marshal.FreeCoTaskMem(strPtr);
+            }
+        }
+    }
+
+    private GitRepositoryStatus RetrieveStatus()
+    {
+        var repoStatus = new GitRepositoryStatus();
+        if (_gitInstalled)
+        {
+            // Options fully explained at https://git-scm.com/docs/git-status
+            // --no-optional-locks : Since this we are essentially running in the background, don't take any optional git locks
+            //                       that could interfere with the user's work. This means calling "status" won't auto-update the
+            //                       index to make future "status" calls faster, but it's better to be unintrusive.
+            // --porcelain=v2      : The v2 gives us nice detailed entries that help us separate ordinary changes from renames, conflicts, and untracked
+            //                       Disclaimer: I'm not sure how far back porcelain=v2 is supported, but I'm pretty sure it's at least 3-4 years.
+            //                       There could be old Git installations that predate it.
+            // -z                  : Terminate filenames and entries with NUL instead of space/LF. This helps us deal with filenames containing spaces.
+            var result = GitExecute.ExecuteGitCommand(_gitDetect.GitConfiguration.ReadInstallPath(), _workingDirectory, "--no-optional-locks status --porcelain=v2 -z");
+            if (result.Status == ProviderOperationStatus.Success && result.Output != null)
+            {
+                var parts = result.Output.Split('\0', StringSplitOptions.RemoveEmptyEntries);
+                for (var i = 0; i < parts.Length; ++i)

```

Comment:
NIT: future devs need to look at a lot of code to read all the ways the line can be parsed.  This could be split up into 4 different lists and each list has their own method for parsing the line.

This way devs can see all the options via the list names.

I'm okay with the code right now.  But I would want to change if a 5th else/if is added.

Reply:
The tricky bit is that for the rename case starting with "2", the entry is actually split across two lines because we get an extra terminator (the format ends with `<path><sep><origPath>`). So we actually need to increment `i` again. It's close to my limit as well, so if it gets any bigger I'll definitely do some cleanup to refactor into more readable chunks.